lib/include/jxl/encode.h

```diff
@@ -264,8 +264,123 @@ JXL_EXPORT JxlEncoderStatus JxlEncoderAddImageFrame(
+ *   decoder.
+ * - "jbrd": JPEG reconstruction box, contains the information required to
+ *   byte-for-byte losslessly recontruct a JPEG-1 image. The JPEG coefficients
+ *   (pixel content) themselves are encoded in the JXL codestream (jxlc or jxlp)
+ *   itself. Exif and XMP metadata will be encoded in Exif and XMP boxes. The
+ *   jbrd box itself contains information such as the app markers of the JPEG-1
+ *   file and everything else required to fit the information together into the
+ *   exact original JPEG file. This box is added automatically by the encoder
+ *   when needed, and only when JPEG reconstruction is used.
+ * - other: other application-specific boxes can be added. Their typename should
+ *   not begin with "jxl" or "JXL" or conflict with other existing typenames.
+ *
+ * Most boxes are automatically added by the encoder and should not be added
+ * with JxlEncoderAddBox. Boxes that one may wish to add with JxlEncoderAddBox
+ * are: Exif and XML (but not when using JPEG reconstruction since if the
+ * JPEG has those, these boxes are already added automatically), jumb, and
+ * application-specific boxes.
+ *
+ * Adding metadata boxes increases the filesize. When adding Exif metadata, the
+ * data must be in sync with what is encoded in the JPEG XL codestream,
+ * specifically the image orientation.
+ *
+ * It is possible to create a codestream without boxes, then what would be in
+ * the jxlc box is written directly to the output
+ *
+ * It is possible to split the codestream across multiple boxes, in that case
+ * multiple boxes of type jxlp are used. This is handled by the encoder when
+ * needed.
+ *
+ * For now metadata boxes can only be added before or after the codestream with
+ * all frames, so using JxlEncoderAddBox is only possible before the first
+ * JxlEncoderAddImageFrame call, and/or after the last JxlEncoderAddImageFrame
+ * call and JxlEncoderCloseInput. Support for adding boxes in-between the
+ * codestream, and/or in-between image frames may be added later, and would
+ * cause the encoder to use jxlp boxes for the codestream.
+ *
+ * @param enc encoder object.
+ * @param type the box type, e.g. "Exif" for EXIF metadata, "XML " for XMP or
+ * IPTC metadata, "jumb" for JUMBF metadata.
+ * @param contents the full contents of the box, for example EXIF
+ * data. For an "Exif" box, the EXIF data must be prepented by a 4-byte tiff
+ * header offset, which may be 4 zero-bytes. The ISO BMFF box header must not
+ * be included, only the contents.
+ * @param size size of the box contents.
+ * @return JXL_ENC_SUCCESS on success, JXL_ENC_ERROR on error, such as when
+ * using this function without JxlEncoderUseContainer, or adding a box type
+ * that would result in an invalid file format.
+ */
+JXL_EXPORT JxlEncoderStatus JxlEncoderAddBox(JxlEncoder* enc, JxlBoxType type,
+                                             const uint8_t* contents,
+                                             size_t size);
```

Comment:
Maybe add an argument `use_brob`? It's fine if it has to be false in the first implementation, but I think that would be the most convenient way to select whether a box is added uncompressed or compressed — an alternative would be to make it an option that is part of the `enc` state, but since you may want to do different things for each box you add, I think it's more convenient to have it as an argument in `JxlEncoderAddBox`.

Reply:
Indeed, that's a good suggestion, done and named it "compress_box"